config/validate/validate.go

```diff
@@ -0,0 +1,174 @@
+		return nil, false
+	}
+	return val, ok
+}
+
+// wrapper for errorutil that handles invalid offsets sanely and resets the reader afterwards
+func posFromOffset(offset int, source io.ReadSeeker) (int, int, string) {
+	if offset == -1 || source == nil {
+		return 0, 0, ""
+	}
+	line, col, highlight := errorutil.HighlightBytePosition(source, int64(offset))
+	source.Seek(0, 0) // Reset the reader to the start so the next call isn't relative to this position
+	return line, col, highlight
+}
+
+func Validate(cfg types.Config, ast json.Node, source io.ReadSeeker) report.Report {
+	v := reflect.ValueOf(cfg)
+	report := validate(v, ast, source)
+	return report
+}
+
+// Validate walks down a struct tree calling AssertValid on every node that implements it, building
+// A report of all the errors, warnings, info, and deprecations it encounters
+func validate(vObj reflect.Value, ast json.Node, source io.ReadSeeker) (r report.Report) {
+	if !vObj.IsValid() {
+		return
+	}
+
+	off := ast.End
+
+	if obj, ok := canCallAssertValid(vObj); ok {
+		if err := obj.AssertValid(); err != nil {
+			line, col, highlight := posFromOffset(off, source)
+			r.Add(report.Entry{
+				Kind:      report.EntryError,
+				Message:   err.Error(),
+				Line:      line,
+				Column:    col,
+				Highlight: highlight,
+			})
+			return //dont recurse on invalid inner nodes, it mostly leads to bogus messages
+		}
+	}
+
+	switch vObj.Kind() {
+	case reflect.Ptr:
+		sub_report := validate(vObj.Elem(), ast, source)
+		sub_report.AddLines(off)
+		r.Merge(sub_report)
+	case reflect.Struct:
+		sub_report := validateStruct(vObj, ast, source)
+		sub_report.AddLines(off)
+		r.Merge(sub_report)
+	case reflect.Slice:
+		for i := 0; i < vObj.Len(); i++ {
+			sub_node := ast
+			if val, ok := ast.Value.([]json.Node); ok {
+				sub_node = val[i]
+			}
+			sub_report := validate(vObj.Index(i), sub_node, source)
+			sub_report.AddLines(off)
+			r.Merge(sub_report)
+		}
+	}
+	return
+}
+
+func ValidateWithoutSource(cfg types.Config) (report report.Report) {
+	return Validate(cfg, json.Node{}, nil)
+}
+
+func validateStruct(vObj reflect.Value, ast json.Node, source io.ReadSeeker) report.Report {
+	off := ast.Start
+	r := report.Report{}
+	keys, ok := ast.Value.(map[string]json.Node)
+	usedKeys := map[string]bool{} // Maintain a set of key's that have been used
+	tags := []string{}            // Maintain a list of all the tags in the struct for fuzzy matching later
+
+	for i := 0; i < vObj.Type().NumField(); i++ {
+		// Default to deepest node if the node's type isn't an object,
+		// such as when a json string actually unmarshal to structs (like with version)
+		sub_node := ast
+		tag := strings.Split(vObj.Type().Field(i).Tag.Get("json"), ",")[0]
+		tags = append(tags, tag) //build list of json key names
+		if ok {
+			sub_node = keys[tag]
+			usedKeys[tag] = true
+		}
+
+		sub_report := validate(vObj.Field(i), sub_node, source)
+		sub_report.AddLines(off)
+		r.Merge(sub_report)
+	}
+	if !ok {
```

Comment:
Because if a JSON type other than an object gets unmarshalled to a struct (this happens with the semvar unmarshaller, also types.Hash), then we still need to walk down its members to validate them. However, since we know it's not an object that was unmarshalled, its pointless to look for unused keys